### PR TITLE
Update Funding Step API operations to use middy validation

### PR DESCRIPTION
**Pull request by zachclark-ccpo**
This replaces API Gateway validation with middy-based validation begun in #302.
API operations for Funding Step of the Portfolio Draft Wizard were changed to perform validation of business rules using middy middleware.

- `createFundingStep`
- `getFundingStep`

### TODO

- [x] Update YAML spec to include empty array validation, minimum value clin validation
- [x] Update fundingStep to include business validation
- [x] Clean up requestValidation (right now its still using a lot of the old FundingStep validation, some of it we don't need)
- [x] Write tests for createFundingStep, getFundingStep
- [x] Update getFundingStep
- [x] Update YAML spec to include date-time validation? This was removed earlier when we were still doing API gateway validation.

See also #381

Ticket: AT-6724

### CODE REVIEW

File: atat_provisioning_wizard_api.yaml
Comment (juliusfitzhugh-ccpo) [inline atat_provisioning_wizard_api.yaml:336]: We might want to talk with the team if we should keep `param-validator` in APIGW or not.

File: infrastructure/lib/load-schema.ts
Comment (juliusfitzhugh-ccpo) [inline infrastructure/lib/load-schema.ts:9]: Maybe I am missing something, but I thought we already made the changes for the `$RefParser` in [PR #372](https://github.com/dod-ccpo/atat-web-api/pull/372/files)?

File: atat_provisioning_wizard_api.yaml
Comment (jasonburkert-ccpo) [inline atat_provisioning_wizard_api.yaml:1031]: If these in-line examples are problematic feel free to remove them.
```suggestion
```

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (juliusfitzhugh-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.ts:77]: Might want to add a `TODO` here as a reminder to change this back to true later, along with a ticket if available.

File: packages/api/portfolioDrafts/funding/getFundingStep.test.ts
Comment (juliusfitzhugh-ccpo) [inline packages/api/portfolioDrafts/funding/getFundingStep.test.ts:37]: I know PR #375 got merged yesterday with the additional `common*MockData` files, so I would suggest adding this to the `commonFundingMockData.ts` with the *good data* after a rebase with those changes.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (juliusfitzhugh-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.test.ts:57]: A comment to open a discussion to help with consistency with tests. After looking at some of the changes from PR #375 I noticed we are going with `() => {}` instead of using `function () {}`. @jasonburkert-ccpo are we moving towards `() => {}` across the codebase? I saw `function () {}` previously in a test and used it because it helped me differentiate the `describe` blocks from the `it` blocks. I just want to make sure we are on the same page so we don't go back and forth between the two.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (juliusfitzhugh-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.test.ts:386]: Everything here and below can probably go in `commonFundingMockData.ts` and then import them.

Comment (zachclark-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:386]: Huh I think I may have accidently overwritten that file after merging, I'll put everything in that file.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (kylelaker-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:57]: Yes -- we should be consistent about using `() => {}` for `describe` and `async () => {}` for `it`

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (kylelaker-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.ts:82]: Just want to point out that this relates to the comments at https://github.com/dod-ccpo/atat-web-api/pull/381#discussion_r742990896

File: packages/api/utils/requestValidation.ts
Comment (kylelaker-ccpo) [inline packages/api/utils/requestValidation.ts:129]: Can we do:
```suggestion
  throw createError(400, "Business rules validation failed", { details: invalidProperties });
```

Looks like `createError` takes a `properties` argument at the end?

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:57]: Agree with Kyle above.  Tried to standardize the describes in last PR.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts:82]: Thanks for linking to that.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (juliusfitzhugh-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:57]: Okay, sounds good. I'll make sure to use `() => {}` moving forward and make changes on my branch as well.

**Review by juliusfitzhugh-ccpo — COMMENTED** [review]
Few small comments, but looks good. Using middy is really coming along well. I like all the additions you added too, they were things that crossed my mind and you added. Nice job!

File: packages/api/utils/requestValidation.ts
Comment (jasonburkert-ccpo) [inline packages/api/utils/requestValidation.ts:50]: Can we use a verb for the function name? Perhaps `validateBusinessRulesForFundingStep`()  

Or, and this might be getting ahead of things, if we used a union type, more simply `validateBusinessRules(ps: PortfolioStep | fs: FundingStep | as: ApplicationStep): ValidationErrorResponse | undefined`

File: packages/api/utils/schemaWrapper.ts
Comment (jasonburkert-ccpo) [inline packages/api/utils/schemaWrapper.ts:9]:
```suggestion
  return {
    type: "object",
    required: ["body"],
    properties: {
      body: schema,
    },
  };
```

File: packages/api/utils/errorHandlingMiddleware.ts
Comment (jasonburkert-ccpo) [inline packages/api/utils/errorHandlingMiddleware.ts:19]: Team: Thoughts on this comment?  `input_validation_errors` was a convention I made up.  In similar discussion with @juliusfitzhugh-ccpo I think I preferred to flatten this and remove `input_validation_errors`.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (jasonburkert-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.test.ts:179]:
```suggestion
```

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:386]: This test needs several adjustments wrt to recently merged #375

File: packages/api/portfolioDrafts/funding/getFundingStep.test.ts
Comment (jasonburkert-ccpo) [inline packages/api/portfolioDrafts/funding/getFundingStep.test.ts:50]:
```suggestion
it("should require path parameter and reject non-UUIDv4 portfolioDraftIds", async () => {
```

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (jasonburkert-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.ts:38]: Do the code level comments add any context?  Is the code self-documenting enough without them?

**Review by jasonburkert-ccpo — CHANGES_REQUESTED** [review]

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (zachclark-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts:38]: That's fair, I just had those comments to help myself organize the file. I'll take them out.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (kylelaker-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:259]: and also, are we sure we want to guarantee ordering of fields?

File: atat_provisioning_wizard_api.yaml
Comment (jasonburkert-ccpo) [reply atat_provisioning_wizard_api.yaml:336]: Per slack we decided to include this in #381, so no need to include here.  We'll use the API GW validation to test for path parameter existence but do all other validation within lambdas.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts
Comment (zachclark-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts:77]: I removed `example: "TaskOrder_9876543210123.pdf"` and was able to get it pass with strict mode.

File: packages/api/utils/schemaWrapper.ts
Comment (zachclark-ccpo) [reply packages/api/utils/schemaWrapper.ts:9]: Smart solutioning, it will be done.

File: packages/api/utils/requestValidation.ts
Comment (zachclark-ccpo) [reply packages/api/utils/requestValidation.ts]: Will change the name, and I like the second idea, probably will roll it into AT-6734.

File: packages/api/utils/errorHandlingMiddleware.ts
Comment (zachclark-ccpo) [reply packages/api/utils/errorHandlingMiddleware.ts:19]: `input_validation_errors` is currently removed, and the thumbs up to @jasonburkert-ccpo comment indicates that we should keep it that way.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts
Comment (zachclark-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:277]: This test uses middy and `mockFundingStepBadData` which is consistent when it comes to the order the error messages come in. We _could_ remove these from this tests, but we can't do it at the `validateClins` level because these checks were removed (middy covers them).

File: packages/api/utils/requestValidation.ts
Comment (kylelaker-ccpo) [inline packages/api/utils/requestValidation.ts:57]:
```suggestion
```

Let's not keep success logs with no related info

File: infrastructure/package-lock.json
Comment (kylelaker-ccpo) [inline infrastructure/package-lock.json:3003]: Not sure I love seeing this pull in an old version in the `package-lock.json`. Can you go ahead and:

```
$ git fetch
$ git checkout origin/develop -- infrastructure/package-lock.json
$ cd infrastructure
$ rm -rf node_modules
$ npm install
```

and just make sure that everything runs okay?

**Review by kylelaker-ccpo — CHANGES_REQUESTED** [review]
small fixes needed but this is looking really good!

File: infrastructure/package-lock.json
Comment (zachclark-ccpo) [reply infrastructure/package-lock.json]: Must've popped in during the merge. I just committed a fix for it.

Comment (jasonburkert-ccpo) [reply infrastructure/package-lock.json]: Fixed in 2119f057bac3b05c00540e396ec74e52db0e7cdd

**Review by kylelaker-ccpo — APPROVED** [review]

**Review by juliusfitzhugh-ccpo — COMMENTED** [review]
Nothing that stands out to me. Nice work!

File: packages/api/utils/errorHandlingMiddleware.ts
Comment (jasonburkert-ccpo) [reply packages/api/utils/errorHandlingMiddleware.ts:19]: This appears to have been addressed (remove input_validation_errors) in `requestValidation.ts`

**Pull request merged**
